src/darwin/Framework/CHIP/ServerEndpoint/MTRServerAccessControl.mm

```
@@ -0,0 +1,193 @@
+    }
+
+    bool GrantSubjectMatchesDescriptor(MTRAccessGrant * grant, const SubjectDescriptor & descriptor)
+    {
+        if (grant.subjectID == nil) {
+            // This is an all-nodes grant for CASE access only.
+            return descriptor.authMode == AuthMode::kCase;
+        }
+
+        NodeId grantSubjectNodeId = grant.subjectID.unsignedLongLongValue;
+        if (IsOperationalNodeId(grantSubjectNodeId)) {
+            return descriptor.authMode == AuthMode::kCase && descriptor.subject == grantSubjectNodeId;
+        }
+
+        if (IsGroupId(grantSubjectNodeId)) {
+            return descriptor.authMode == AuthMode::kGroup && descriptor.subject == grantSubjectNodeId;
+        }
+
+        if (IsCASEAuthTag(grantSubjectNodeId)) {
+            return descriptor.cats.CheckSubjectAgainstCATs(grantSubjectNodeId);
+        }
+
+        MTR_LOG_ERROR("Unexpected grant subject: 0x%llx", grantSubjectNodeId);
+        return false;
+    }
+};
+
+struct ControllerAccessControl {
+    DeviceTypeResolver mDeviceTypeResolver;
+    AccessControlDelegate mDelegate;
+    ControllerAccessControl() { GetAccessControl().Init(&mDelegate, mDeviceTypeResolver); }
+};
+
+Global<ControllerAccessControl> gControllerAccessControl;
+
+} // anonymous namespace
+
+int MatterGetAccessPrivilegeForReadEvent(ClusterId cluster, EventId event)
+{
+    // We don't support any event bits yet.
+    return kMatterAccessPrivilegeAdminister;
+}
+
+int MatterGetAccessPrivilegeForInvokeCommand(ClusterId cluster, CommandId command)
+{
+    // For now we only have OTA, which uses Operate.
+    return kMatterAccessPrivilegeOperate;
+}
+
+int MatterGetAccessPrivilegeForReadAttribute(ClusterId cluster, AttributeId attribute)
+{
+    NSNumber * _Nullable neededPrivilege = [[MTRDeviceControllerFactory sharedInstance] neededReadPrivilegeForClusterID:@(cluster) attributeID:@(attribute)];
+    if (neededPrivilege == nil) {
+        // Default is View.
+        return kMatterAccessPrivilegeView;
```

Where does this default come from? Maybe by your logic for unknown values this should also be Admin?

---

Default comes from spec, sort of.  But you are right; probably better to fail closed here.  Will fix.

---

Done.